src/services/RecordStreamProcessor.ts

```diff
@@ -0,0 +1,110 @@
+    }
+    return RecordStreamProcessor.instance;
+  }
+
+  async processRequest(sessionEvent: any): Promise<void> {
+
+    let sessionEventData: any = ExtSessionEvent.parseRequest(JSON.stringify(sessionEvent));
+
+    this.logger.appendKeys({ govuk_signin_journey_id: sessionEventData.clientSessionId });
+
+    if (sessionEventData.notified) {
+      this.handleNotifiedUser();
+    }
+
+    try {
+      this.validationHelper.validateSessionEventFields(sessionEventData);
+    } catch (error: any) {
+      this.handleValidationFailure(error);
+    }
+
+    let emailType = Constants.VIST_PO_EMAIL_DYNAMIC;
+    if (!sessionEventData.documentUploadedOn || !(sessionEventData.documentUploadedOn > 0)) {
+      const result = this.handleDocumentUploadedOnNotSet(sessionEventData);
+      emailType = result.emailType;
+      sessionEventData = result.sessionEventData;
+    }
+
+    const data = await this.validationHelper.validateSessionEvent(sessionEventData, emailType, this.logger);
+
+    await this.sendEmailMessageToGovNotify(data.sessionEvent, data.emailType);
+
+    try {
+      await this.iprService.setNotifiedFlag(data.sessionEvent.userId, true);
+    } catch (error: any) {
+      throw new AppError(HttpCodesEnum.SERVER_ERROR, error.message);
+    }
+
+  }
+
+  private handleNotifiedUser(): void {
+    this.logger.warn("User is already notified for this session event.", { messageCode: MessageCodes.USER_ALREADY_NOTIFIED });
+    throw new AppError(HttpCodesEnum.SERVER_ERROR, "User is already notified for this session event.");
+  }
+
+  private handleValidationFailure(error: any): void {
+    this.logger.warn(error.message, { messageCode: MessageCodes.MISSING_MANDATORY_FIELDS_IN_SESSION_EVENT });
+    throw new AppError(HttpCodesEnum.SERVER_ERROR, error.message);
+  }
+
+  private handleDocumentUploadedOnNotSet(eventData: any): { emailType: string; sessionEventData: SessionEvent; } {
+    this.logger.info({ message: "documentUploadedOn is not yet populated, sending the old template email." });
+    const emailType = Constants.VIST_PO_EMAIL_STATIC;
+    const sessionEventData = new SessionEvent(eventData);
+    return { emailType, sessionEventData }
+  }
+
+  private async sendEmailMessageToGovNotify(sessionEvent: ExtSessionEvent | SessionEvent, emailType: string): Promise<void> {
+    try {
+      this.logger.info({ message: `Trying to send ${emailType} type message to GovNotify handler` });
+      const nameParts = personalIdentityUtils.getNames(sessionEvent.nameParts);
+      await this.iprService.sendToGovNotify(buildGovNotifyEventFields(nameParts, sessionEvent, emailType, this.logger));
+      this.logger.info({ message: `Sent ${emailType} message to GovNotify handler` });
+    } catch (error) {
+      this.handleGovNotifyError(error, emailType);
+    }
+  }
+
+  private handleGovNotifyError(error: any, emailType: string): void {
+    this.logger.error("FAILED_TO_WRITE_GOV_NOTIFY", {
```

Comment:
This log is a bit of a weird format - could we change it to contain `message`, `error` and `messageCode`?